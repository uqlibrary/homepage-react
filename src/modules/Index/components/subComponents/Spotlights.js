import React from 'react';
import { PropTypes } from 'prop-types';
import {
    CarouselProvider,
    Slider,
    Slide,
    ButtonBack,
    ButtonNext,
    ButtonPlay,
    DotGroup,
    Image,
} from 'pure-react-carousel';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import ChevronRightIcon from '@material-ui/icons/ChevronRight';
import PlayArrowIcon from '@material-ui/icons/PlayArrow';
import PauseIcon from '@material-ui/icons/Pause';
const defaultSlide = require('../../../../../public/images/Welcome_Spotlight.jpg');

const Spotlights = ({ spotlights, spotlightsLoading, account }) => {
    const totalSlides = spotlights && spotlights.length;
    const slides =
        spotlights && spotlights.length > 0
            ? spotlights.map((item, index) => {
                  return {
                      index,
                      src: item.img_url,
                      alt: item.img_alt,
                      aria: item.title,
                      link: item.url,
                  };
              })
            : [];
    if (slides.length > 0 && !spotlightsLoading) {
        return (
            <div
                data-testid="spotlights"
                style={{ height: '100%', position: 'relative', borderRadius: 4, overflow: 'hidden' }}
                role="region"
                aria-label="UQ Spotlights carousel"
            >
                <CarouselProvider
                    visibleSlides={1}
                    totalSlides={totalSlides}
                    step={1}
                    naturalSlideWidth={1967}
                    naturalSlideHeight={721}
                    isPlaying={!account}
                    interval={10000}
                    style={{ height: '100%' }}
                >
                    {totalSlides > 1 && (
                        <div
                            style={{
                                width: 36,
                                height: '100%',
                                position: 'absolute',
                                top: 4,
                                left: 4,
                                zIndex: 101,
                            }}
                        >
                            <ButtonBack>
                                <ChevronLeftIcon />
                            </ButtonBack>
                        </div>
                    )}
                    {totalSlides > 1 && (
                        <div
                            style={{
                                width: 36,
                                height: '100%',
                                position: 'absolute',
                                top: 4,
                                right: 0,
                                zIndex: 101,
                            }}
                        >
                            <ButtonNext>
                                <ChevronRightIcon />
                            </ButtonNext>
                        </div>
                    )}
                    {totalSlides > 1 && (
                        <div
                            style={{
                                width: 36,
                                position: 'absolute',
                                top: 4,
                                left: 32,
                                zIndex: 101,
                            }}
                        >
                            <ButtonPlay childrenPlaying={<PauseIcon />} childrenPaused={<PlayArrowIcon />} />
                        </div>
                    )}
                    {totalSlides > 1 && (
                        <div
                            style={{
                                width: 200,
                                marginLeft: -100,
                                position: 'absolute',
                                top: 0,
                                left: '50%',
                                zIndex: 101,
                                textAlign: 'center',
                            }}
                        >
                            <DotGroup />
                        </div>
                    )}
                    <div
                        style={{
                            zIndex: 99,
                        }}
                    >
                        <Slider aria-label="UQ Spotlights slider" trayTag="div">
                            {slides.map((item, index) => {
                                return (
                                    <Slide tag="div" index={item.index} key={index}>
                                        <a href={item.link} aria-label={item.aria}>
                                            <Image
                                                src={item.src}
                                                alt={item.alt}
                                                style={{ width: '100%' }}
                                                aria-label={item.aria}
                                            />
                                        </a>
                                    </Slide>
                                );
                            })}
                        </Slider>
                    </div>
                </CarouselProvider>
            </div>
        );
    }
    return null;
};

Spotlights.propTypes = {
    spotlights: PropTypes.array,
    spotlightsLoading: PropTypes.bool,
    defaultSlide: PropTypes.string,
    account: PropTypes.object,
};

Spotlights.defaultProps = {
    spotlightsLoading: false,
    defaultSlide: defaultSlide,
};

export default Spotlights;
