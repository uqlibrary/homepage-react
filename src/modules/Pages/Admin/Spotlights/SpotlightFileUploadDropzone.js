import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { useDropzone } from 'react-dropzone';

import Grid from '@material-ui/core/Grid';
import DeleteIcon from '@material-ui/icons/Delete';
import IconButton from '@material-ui/core/IconButton';
import { default as locale } from './spotlightsadmin.locale';

const thumbsContainer = {
    borderRadius: 2,
    border: '1px solid #eaeaea',
    marginBottom: 8,
    marginRight: 8,
    height: 200,
    padding: 4,
    boxSizing: 'border-box',
    minHeight: 320,
};

const thumbInner = {
    display: 'flex',
    minWidth: 0,
    overflow: 'hidden',
};

const thumbImg = {
    maxWidth: '100%',
    maxHeight: 200,
};

const deleteButton = {
    width: 80,
    height: 80,
};

const warningDimensions = {
    color: 'red',
    fontWeight: 'bold',
};
const okDimensions = {
    // it doesnt like not having a style provided in the else
    color: 'inherit',
};

const dimensionBox = {
    padding: '1rem',
};

export function SpotlightFileUploadDropzone({ onAddFile, onClearFile }) {
    const [files, setFiles] = useState([]);
    const [imageWidth, setImageWidth] = useState(0);
    const [imageHeight, setImageHeight] = useState(0);

    const getDimensions = acceptedFiles => {
        if (!acceptedFiles || acceptedFiles.ength === 0) {
            return;
        }
        const url = URL.createObjectURL(acceptedFiles[0]);
        const img = new Image();

        img.onload = () => {
            setImageWidth(img.width);
            setImageHeight(img.height);

            URL.revokeObjectURL(img.src);
        };
        img.src = url;
    };

    const { getRootProps, getInputProps } = useDropzone({
        accept: 'image/*',
        onDrop: acceptedFiles => {
            setFiles(
                acceptedFiles.map(file =>
                    Object.assign(file, {
                        preview: URL.createObjectURL(file),
                    }),
                ),
            );
            onAddFile(acceptedFiles);
            getDimensions(acceptedFiles);
        },
        maxFiles: 1,
    });

    useEffect(
        () => () => {
            // Make sure to revoke the data uris to avoid memory leaks
            files.forEach(file => URL.revokeObjectURL(file.preview));
        },
        [files],
    );

    const removeUpload = () => {
        console.log('remove uploaded file ', files);
        files.forEach(file => URL.revokeObjectURL(file.preview));
        setFiles([]);

        onClearFile();
    };

    return (
        <section className="container" data-testid="spotlights-form-upload-dropzone">
            {!files || files.length === 0 ? (
                <div
                    {...getRootProps({ className: 'dropzone' })}
                    style={{
                        border: 'thin solid black',
                        backgroundColor: 'lightgrey',
                        padding: '1rem',
                        cursor: 'pointer',
                    }}
                >
                    <input data-testid="dropzone-dragarea" {...getInputProps()} />
                    <p>{locale.form.labels.dragareaInstructions}</p>
                </div>
            ) : (
                <Grid container data-testid="dropzone-preview" style={thumbsContainer}>
                    <Grid item xs={12}>
                        <h3>Preview:</h3>
                    </Grid>
                    <Grid item xs={10}>
                        {files.map(file => (
                            <div style={thumbInner} key={file.name}>
                                <img alt="preview of uploaded spotlight file" src={file.preview} style={thumbImg} />
                            </div>
                        ))}
                    </Grid>
                    <Grid item xs={2} align="center">
                        <IconButton style={deleteButton} onClick={removeUpload} title={locale.form.tooltips.deleteIcon}>
                            <DeleteIcon />
                        </IconButton>
                    </Grid>
                    <Grid item xs={12}>
                        {files.map(file => (
                            <div key={`${file.name}-dimensions`}>
                                {imageWidth > 0 && imageHeight > 0 && (
                                    <div style={dimensionBox} data-testid="dropzone-dimension-warning">
                                        <div
                                            style={
                                                imageWidth > locale.form.image.maxWidth ||
                                                imageHeight > locale.form.image.maxHeight
                                                    ? warningDimensions
                                                    : okDimensions
                                            }
                                        >
                                            Dimensions:{' '}
                                            <strong>
                                                {imageWidth}px x {imageHeight}px
                                            </strong>
                                        </div>
                                        {locale.form.image.dimensionsNotification}: {locale.form.image.maxWidth}px x{' '}
                                        {locale.form.image.maxHeight}px
                                    </div>
                                )}
                            </div>
                        ))}
                    </Grid>
                </Grid>
            )}
        </section>
    );
}

SpotlightFileUploadDropzone.propTypes = {
    onAddFile: PropTypes.func,
    onClearFile: PropTypes.func,
};
