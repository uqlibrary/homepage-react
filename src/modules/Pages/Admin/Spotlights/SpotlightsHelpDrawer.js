import React from 'react';
import PropTypes from 'prop-types';

import SimpleBackdrop from '@material-ui/core/Backdrop';
import Button from '@material-ui/core/Button';
import Fade from '@material-ui/core/Fade';
import { makeStyles } from '@material-ui/styles';
import Drawer from '@material-ui/core/Drawer';

const useStyles = makeStyles(
    theme => ({
        drawer: {
            overflowY: 'scroll',
            '& p': {
                marginBlockStart: 0,
                marginBlockEnd: '1em',
            },
            '& li': {
                marginBlockStart: 0,
                marginBlockEnd: '1em',
            },
            '& dt': {
                fontStyle: 'italic',
            },
        },
        paper: {
            backgroundColor: theme.palette.background.paper,
            padding: theme.spacing(2, 4, 3),
            width: 500,
            [theme.breakpoints.down('sm')]: {
                width: 200,
            },
        },
    }),
    { withTheme: true },
);
export const SpotlightsHelpDrawer = ({ open, helpContent, closeHelpLightbox }) => {
    const classes = useStyles();

    return (
        <Drawer
            anchor="right"
            className={classes.drawer}
            open={open}
            onClose={closeHelpLightbox}
            closeAfterTransition
            BackdropComponent={SimpleBackdrop}
            BackdropProps={{
                timeout: 500,
            }}
        >
            <Fade in={open}>
                <div className={classes.paper}>
                    <h2>{helpContent?.title || /* istanbul ignore next */ 'TBA'}</h2>
                    <div>{helpContent?.text || /* istanbul ignore next */ ''}</div>
                    <div>
                        <Button
                            variant="contained"
                            color="primary"
                            className={classes.button}
                            children={helpContent?.buttonLabel || 'Close'}
                            onClick={closeHelpLightbox}
                        />
                    </div>
                </div>
            </Fade>
        </Drawer>
    );
};

SpotlightsHelpDrawer.propTypes = {
    // actions: PropTypes.any,
    helpContent: PropTypes.any,
    // helpButtonLabel: PropTypes.string,
    // history: PropTypes.object,
    // showAddButton: PropTypes.bool,
    open: PropTypes.any,
    setHelpLightboxOpen: PropTypes.any,
    closeHelpLightbox: PropTypes.any,
};

// SpotlightsHelpDrawer.defaultProps = {
//     helpButtonLabel: 'Help',
//     showAddButton: false,
// };
