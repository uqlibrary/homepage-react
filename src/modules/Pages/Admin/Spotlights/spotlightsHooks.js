/* eslint-disable react-hooks/rules-of-hooks */
import { useState, useCallback } from 'react';

export const useViewByHistoryLightboxState = () => {
    const [isViewByHistoryLightboxOpen, setViewByHistoryLightboxOpen] = useState(false);

    const handleViewByHistoryLightboxOpen = useCallback(() => {
        setViewByHistoryLightboxOpen(true);
    }, []);

    const handleViewByHistoryLightboxClose = useCallback(() => {
        setViewByHistoryLightboxOpen(false);
    }, []);

    return [isViewByHistoryLightboxOpen, handleViewByHistoryLightboxOpen, handleViewByHistoryLightboxClose];
};
