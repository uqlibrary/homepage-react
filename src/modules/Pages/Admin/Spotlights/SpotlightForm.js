import React, { Fragment, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import Button from '@material-ui/core/Button';
import Checkbox from '@material-ui/core/Checkbox';
import FormControl from '@material-ui/core/FormControl';
import Grid from '@material-ui/core/Grid';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import { makeStyles } from '@material-ui/styles';
import { KeyboardDateTimePicker } from '@material-ui/pickers';

import { ConfirmationBox } from 'modules/SharedComponents/Toolbox/ConfirmDialogBox';
import { InlineLoader } from 'modules/SharedComponents/Toolbox/Loaders';
import { SpotlightUploader } from 'modules/Pages/Admin/Spotlights/SpotlightUploader';
import { default as locale } from './spotlightsadmin.locale';
// import { formatDate, getTimeEndOfDayFormatted, getTimeNowFormatted } from './spotlighthelpers';
import { formatDate } from './spotlighthelpers';

import { useConfirmationState } from 'hooks';

const moment = require('moment');

const useStyles = makeStyles(() => ({
    saveButton: {
        '&:disabled': {
            color: 'rgba(0, 0, 0, 0.26)',
            boxShadow: 'none',
            backgroundColor: 'rgba(0, 0, 0, 0.12)',
        },
    },
}));

export const SpotlightForm = ({
    actions,
    spotlightsLoading,
    spotlightResponse,
    spotlightStatus,
    defaults,
    spotlightError,
    publicFileUploading,
    publicFileUploadError,
    publicFileUploadResult,
    history,
}) => {
    console.log('uploadPublicFile: publicFileUploading = ', publicFileUploading);
    console.log('uploadPublicFile: publicFileUploadError = ', publicFileUploadError);
    console.log('uploadPublicFile: publicFileUploadResult = ', publicFileUploadResult);
    // !!publicFileUploadResult
    //     ? console.log('uploadPublicFile: publicFileUploadResult join = ', publicFileUploadResult.join(' ').trim())
    //     : console.log('no join publicFileUploadResult = ', publicFileUploadResult);
    const classes = useStyles();

    const [isOpen, showConfirmation, hideConfirmation] = useConfirmationState();
    // const [isOpenUploadFile, showErrorUploadFile, hideConfirmationUploadFile] = useConfirmationState();

    const [isFormValid, setFormValidity] = useState(false); // enable-disable the save button
    const [uploadedFiles, setUploadedFiles] = useState(null);

    console.log('defaults = ', defaults);
    const [values, setValues] = useState({
        // the data displayed in the form
        ...defaults,
        start: defaults.startDateDefault,
        end: defaults.endDateDefault,
    });

    const isValidTitle = title => {
        return title.length > 0;
    };

    const isValidUrl = testurl => {
        if (!!testurl && testurl.length < 'http://x.co'.length) {
            // minimum possible url
            return false;
        }
        try {
            const url = new URL(testurl);
            if (url.hostname.length < 'x.co'.length) {
                return false;
            }
        } catch (_) {
            return false;
        }
        // while technically an url doesn't need a TLD - in practice it does
        if (!testurl.includes('.')) {
            return false;
        }
        return true;
    };

    function isValidStartDate(param, startDate) {
        const momentStartDate = new moment(startDate);
        const momentStartDateFormatted = momentStartDate.format('YYYY-MM-DDTHH:mm');
        return (
            startDate !== '' && momentStartDateFormatted >= defaults.startDateDefault && !!moment(startDate).isValid()
        );
    }

    function isInvalidStartDate(param, startDate) {
        return !isValidStartDate(param, startDate);
    }

    function isInvalidEndDate(endDate, startDate) {
        return (endDate < startDate && startDate !== '') || !moment(endDate).isValid();
    }

    const validateValues = currentValues => {
        console.log('validateValues: currentValues = ', currentValues);
        const isValid =
            !spotlightsLoading &&
            !isInvalidStartDate(null, currentValues.start) &&
            !isInvalidEndDate(currentValues.end, currentValues.start) &&
            !!isValidTitle(currentValues.title) &&
            // currentValues.img_alt.length > 0 && // set to title during save if blank
            // !!currentValues.localfilename &&
            // currentValues.localfilename.length > 0 &&
            !!currentValues.uploadedFile &&
            // currentValues.fileDetails.length > 0 &&
            !!currentValues.url &&
            currentValues.url.length > 0 &&
            isValidUrl(currentValues.url);

        console.log('validateValues: isValid = ', isValid, currentValues);
        // console.log(
        //     'validateValues: isInvalidStartDate(',
        //     currentValues.start,
        //     ') = ',
        //     isInvalidStartDate(currentValues.start),
        // );
        // console.log('validateValues: currentValues.end = ', currentValues.end);
        // console.log(
        //     'validateValues: isInvalidStartDate(',
        //     currentValues.end,
        //     ') = ',
        //     isInvalidStartDate(currentValues.end),
        // );
        // console.log('validateValues: spotlightsLoading = ', spotlightsLoading);
        // console.log('validateValues: isValidTitle(', currentValues.title, ') = ', isValidTitle(currentValues.title));
        // console.log('validateValues: isValidUrl(', currentValues.url, ') = ', isValidUrl(currentValues.url));
        // console.log('validateValues: currentValues.img_alt = ', currentValues.img_alt);
        // console.log('validateValues: currentValues = ', currentValues);

        return isValid;
    };

    useEffect(() => {
        if (!!defaults && defaults.type === 'clone') {
            setFormValidity(validateValues(defaults));
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    useEffect(() => {
        // console.log('uploadedFiles set to ', uploadedFiles);
        console.log('values have been changed to: ', values);
        setFormValidity(validateValues(values));
    }, [values]);

    useEffect(() => {
        if (!!spotlightResponse && !!spotlightResponse.id && spotlightStatus === 'saved') {
            showConfirmation();
        }
    }, [showConfirmation, spotlightResponse, spotlightStatus]);

    useEffect(() => {
        if (!!spotlightError || spotlightStatus === 'error') {
            showConfirmation();
        }
    }, [showConfirmation, spotlightError, spotlightStatus]);

    useEffect(() => {
        if (!!publicFileUploadError) {
            showConfirmation();
        }
    }, [showConfirmation, publicFileUploadError]);

    const addAriaLabelToMuiDatePickerButton = (idDiv, ariaLabel) => {
        const theBlock = document.getElementById(idDiv);
        const theButton = !!theBlock && theBlock.parentNode.querySelector('button');
        !!theButton && !theButton.hasAttribute('aria-label') && theButton.setAttribute('aria-label', ariaLabel);
    };

    const runAfterRender = () => {
        // component doesnt allow pass of aria-label to the button, and we have 2, so they need distinct labels
        addAriaLabelToMuiDatePickerButton('admin-spotlights-form-start-date-label', 'Select publish date-time');
        addAriaLabelToMuiDatePickerButton('admin-spotlights-form-end-date-label', 'Select unpublish date-time');
    };

    const clearForm = () => {
        setValues(defaults);
        // setValues({
        //     ['title']: '',
        //     ['url']: '',
        //     ['start']: defaults.startDateDefault,
        //     ['end']: defaults.endDateDefault,
        //     ['urgent']: false,
        //     ['img_alt']: '',
        //     ['img_url']: '',
        //     ['weight']: 0,
        //     ['active']: 0,
        //     startDateDefault: getTimeNowFormatted(),
        //     endDateDefault: getTimeEndOfDayFormatted(),
        // });
    };

    const navigateToListPage = () => {
        clearForm();

        actions.clearSpotlights(); // force the list page to reload after save

        actions.clearASpotlight(); // make the form clear for the next use

        history.push('/admin/spotlights');

        const topOfPage = document.getElementById('StandardPage');
        !!topOfPage && topOfPage.scrollIntoView();
    };

    const reloadClonePage = () => {
        setValues(defaults);

        const topOfPage = document.getElementById('StandardPage');
        !!topOfPage && topOfPage.scrollIntoView();
    };

    const handleSpotlightCreation = newValues => {
        console.log('handleSpotlightCreation 1: newValues = ', newValues);
        if (defaults.type === 'add') {
            console.log('handleSpotlightCreation: uploadedFiles = ', uploadedFiles);
            // only 1 file may be uploaded, but it comes in an array
            // const uploadedFile = !!uploadedFiles && uploadedFiles.lengh > 0 && uploadedFiles.shift();
            // newValues.uploadedFile = uploadedFile;
            console.log('handleSpotlightCreation 2: newValues = ', newValues);
            // !!uploadedFile && actions.createSpotlightWithFile(newValues);
            actions.createSpotlightWithFile(newValues);
        } else {
            // newValues.img_url should be supplied by the form, because we preview the image in there
            actions.createSpotlight(newValues);
        }
    };

    const saveSpotlight = () => {
        console.log('saveSpotlight: values = ', values);
        const newValues = {
            id: defaults.type !== 'add' ? values.id : null,
            start: formatDate(values.start),
            end: formatDate(values.end),
            title: values.title,
            url: values.url,
            // img_url: values.img_url,
            img_alt: values.img_alt || values.title,
            weight: values.weight,
            active: !!values.active ? 1 : 0,
        };

        console.log('saveSpotlight defaults.type = ', defaults.type);
        console.log(
            defaults.type === 'edit'
                ? 'saveSpotlight actions.saveSpotlightChange: newValues = '
                : 'saveSpotlight handleSpotlightCreation: newValues = ',
        );
        console.log('saveSpotlight: newValues = ', newValues);
        defaults.type === 'edit' ? actions.saveSpotlightChange(newValues) : handleSpotlightCreation(newValues);

        // force to the top of the page, because otherwise it looks a bit weird
        window.scrollTo({
            top: 0,
            left: 0,
            behavior: 'smooth',
        });
    };

    const handleChange = prop => event => {
        let newValue;
        if (['start', 'end'].includes(prop)) {
            newValue = event.format('YYYY/MM/DD hh:mm a');
        } else {
            newValue = !!event.target.value ? event.target.value : event.target.checked;
            if (prop === 'active') {
                newValue = !!newValue ? 1 : 0;
            }
        }
        console.log('handleChange ', prop, ': newValue = ', newValue);
        setValues({ ...values, [prop]: newValue });

        console.log('handleChange values now = ', values);
        setFormValidity(validateValues({ ...values, [prop]: newValue }));
    };

    const errorLocale = {
        ...locale.form.add.addSpotlightError,
        confirmationTitle: `An error occurred: ${spotlightError}`,
    };

    const uploadErrorLocale = () => {
        console.log('uploadErrorLocale: publicFileUploadResult = ', publicFileUploadResult);
        const errorMessage = (!!publicFileUploadResult && publicFileUploadResult[0]) || '';
        return {
            ...locale.form.upload.uploadError,
            // confirmationTitle: 'An error occurred during the upload',
            confirmationTitle: `An error occurred during the upload${!!errorMessage ? ': ' + errorMessage.trim() : ''}`,
        };
    };

    const handleConfirmation = () => {
        if (defaults.type === 'edit') {
            // the action on edit page is always 'return to list'
            navigateToListPage();
        } else if (!!spotlightError) {
            // On error on creation, the button just closes the notification dialog,
            // allowing the user to correct and try again
            hideConfirmation(); // form remains loaded
        } else {
            clearForm();
        }
    };

    if (!!publicFileUploading) {
        return (
            <Grid
                item
                xs={'auto'}
                style={{
                    width: 80,
                    marginRight: 20,
                    marginBottom: 6,
                    opacity: 0.3,
                    height: 200, // default to some space for the blocks
                }}
            >
                <InlineLoader message="Loading" />
            </Grid>
        );
    }

    const handleUploadedFiles = files => {
        console.log('handleUploadedFiles files = ', files);
        const file = !!files && files.length > 0 && files.shift();
        console.log('handleUploadedFiles file = ', file);

        setUploadedFiles(file);

        // const filename = !!file && file.name;
        // console.log('handleUploadedFiles filename = ', filename);
        setValues({ ...values, ['uploadedFile']: file });

        console.log('handleUploadedFiles values now = ', values);
        setFormValidity(validateValues({ ...values, ['uploadedFile']: file }));

        setTimeout(() => {
            console.log('handleUploadedFiles setTimeout: uploadedFiles = ', uploadedFiles);
        }, 1000);
    };

    const clearUploadedFile = () => {
        setFormValidity(validateValues({ ...values, ['uploadedFile']: '' }));
    };

    return (
        <Fragment>
            <form onLoad={runAfterRender()}>
                {spotlightStatus === 'error' && (
                    <ConfirmationBox
                        actionButtonColor="primary"
                        actionButtonVariant="contained"
                        confirmationBoxId="spotlight-error"
                        onAction={() =>
                            spotlightError === 'The requested page could not be found.' && navigateToListPage()
                        }
                        onClose={hideConfirmation}
                        hideCancelButton
                        isOpen={isOpen}
                        locale={errorLocale}
                    />
                )}
                {spotlightStatus !== 'error' && defaults.type === 'edit' && (
                    <ConfirmationBox
                        actionButtonColor="primary"
                        actionButtonVariant="contained"
                        confirmationBoxId="spotlight-edit-save-succeeded"
                        onAction={handleConfirmation}
                        onClose={hideConfirmation}
                        hideCancelButton
                        isOpen={isOpen}
                        locale={locale.form.edit.editSpotlightConfirmation}
                    />
                )}
                {spotlightStatus !== 'error' && !publicFileUploadError && defaults.type === 'add' && (
                    <ConfirmationBox
                        actionButtonColor="secondary"
                        actionButtonVariant="contained"
                        confirmationBoxId="spotlight-add-save-succeeded"
                        onAction={handleConfirmation}
                        onClose={hideConfirmation}
                        onCancelAction={() => navigateToListPage()}
                        isOpen={isOpen}
                        locale={locale.form.add.addSpotlightConfirmation}
                    />
                )}
                {spotlightStatus !== 'error' && defaults.type === 'clone' && (
                    <ConfirmationBox
                        actionButtonColor="secondary"
                        actionButtonVariant="contained"
                        confirmationBoxId="spotlight-clone-save-succeeded"
                        onClose={hideConfirmation}
                        onAction={() => reloadClonePage()}
                        isOpen={isOpen}
                        locale={locale.form.clone.cloneSpotlightConfirmation}
                        onCancelAction={() => navigateToListPage()}
                    />
                )}
                {!!publicFileUploadError && (
                    <ConfirmationBox
                        actionButtonColor="primary"
                        actionButtonVariant="contained"
                        confirmationBoxId="spotlight-file-upload-failed"
                        onClose={hideConfirmation}
                        onAction={() => hideConfirmation()}
                        isOpen={isOpen}
                        locale={uploadErrorLocale()}
                        hideCancelButton
                    />
                )}
                <Grid container spacing={2}>
                    <Grid item xs={12}>
                        <FormControl fullWidth title={locale.form.tooltips.ariaTitle}>
                            <InputLabel htmlFor="spotlightTitle">Title - visible to assitive technology *</InputLabel>
                            <Input
                                id="spotlightTitle"
                                data-testid="admin-spotlights-form-title"
                                error={!isValidTitle(values.title)}
                                inputProps={{ maxLength: 100 }}
                                onChange={handleChange('title')}
                                required
                                value={values.title}
                            />
                        </FormControl>
                    </Grid>
                </Grid>
                <Grid container spacing={2}>
                    <Grid item xs={12}>
                        <FormControl fullWidth title={locale.form.tooltips.tooltipMouseover}>
                            <InputLabel htmlFor="spotlightTooltip">{locale.form.tooltips.tooltipInField}</InputLabel>
                            <Input
                                id="spotlightTooltip"
                                data-testid="admin-spotlights-form-tooltip"
                                value={values.img_alt}
                                onChange={handleChange('img_alt')}
                                inputProps={{ maxLength: 100 }}
                            />
                        </FormControl>
                    </Grid>
                </Grid>
                <Grid container spacing={2}>
                    <Grid item xs={12}>
                        <FormControl fullWidth title={locale.form.tooltips.link}>
                            <InputLabel htmlFor="linkUrl">Spotlight link *</InputLabel>
                            <Input
                                type="url"
                                id="linkUrl"
                                data-testid="admin-spotlights-form-link-url"
                                value={values.url}
                                onChange={handleChange('url')}
                                error={!isValidUrl(values.url)}
                            />
                        </FormControl>
                    </Grid>
                </Grid>
                <Grid container spacing={2} style={{ marginTop: 12 }}>
                    <Grid item md={5} xs={12}>
                        <KeyboardDateTimePicker
                            id="admin-spotlights-form-start-date"
                            data-testid="admin-spotlights-form-start-date"
                            value={values.start}
                            label="Date published"
                            onChange={handleChange('start')}
                            minDate={defaults.minimumDate}
                            format="DD/MM/YYYY HH:mm a"
                            showTodayButton
                            autoOk
                        />
                    </Grid>
                    <Grid item md={5} xs={12}>
                        <KeyboardDateTimePicker
                            id="admin-spotlights-form-end-date"
                            data-testid="admin-spotlights-form-end-date"
                            label="Date unpublished"
                            onChange={handleChange('end')}
                            value={values.end}
                            minDate={values.start}
                            format="DD/MM/YYYY HH:mm a"
                            autoOk
                        />
                    </Grid>
                </Grid>
                <Grid container spacing={2} style={{ marginTop: '1rem' }}>
                    <Grid item xs={10} align="left">
                        <SpotlightUploader onAddFile={handleUploadedFiles} onClearFile={clearUploadedFile} />
                    </Grid>
                </Grid>
                <Grid container spacing={2} style={{ marginTop: '1rem' }}>
                    <Grid item xs={3} align="left">
                        <InputLabel
                            style={{ color: 'rgba(0, 0, 0, 0.87)' }}
                            title={locale.form.tooltips.publishcheckbox}
                        >
                            <Checkbox
                                checked={values.active === 1}
                                data-testid="admin-spotlights-form-checkbox-published"
                                onChange={handleChange('active')}
                                className={classes.checkbox}
                            />
                            Published?
                        </InputLabel>
                    </Grid>
                </Grid>
                <Grid container spacing={2} style={{ marginTop: '1rem' }}>
                    <Grid item xs={3} align="left">
                        <Button
                            color="secondary"
                            children="Cancel"
                            data-testid="admin-spotlights-form-button-cancel"
                            onClick={() => navigateToListPage()}
                            variant="contained"
                        />
                    </Grid>
                    <Grid item xs={9} align="right">
                        <Button
                            color="primary"
                            data-testid="admin-spotlights-form-button-save"
                            variant="contained"
                            children={defaults.type === 'edit' ? 'Save' : 'Create'}
                            disabled={!isFormValid}
                            onClick={saveSpotlight}
                            className={classes.saveButton}
                        />
                    </Grid>
                </Grid>
            </form>
        </Fragment>
    );
};

SpotlightForm.propTypes = {
    actions: PropTypes.any,
    publicFileUploading: PropTypes.any,
    publicFileUploadError: PropTypes.any,
    publicFileUploadResult: PropTypes.any,
    spotlightResponse: PropTypes.any,
    spotlightError: PropTypes.any,
    spotlightsLoading: PropTypes.any,
    spotlightStatus: PropTypes.any,
    defaults: PropTypes.object,
    history: PropTypes.object,
};

SpotlightForm.defaultProps = {
    publicFileUploading: false, // whether a file is currently being uploaded. Only done by Add, other defaults false
    publicFileUploadError: false,
    publicFileUploadResult: false,
};

export default SpotlightForm;
