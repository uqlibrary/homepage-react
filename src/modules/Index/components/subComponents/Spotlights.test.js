import React from 'react';
import { rtlRender } from 'test-utils';
import Spotlights from './Spotlights';

function setup(testProps = {}) {
    const props = {
        spotlights: [],
        account: {},
    };
    return rtlRender(<Spotlights {...props} {...testProps} />);
}

describe('Spotlights', () => {
    it('default values render correctly', () => {
        const { container } = setup();
        expect(container).toMatchSnapshot();
    });
});