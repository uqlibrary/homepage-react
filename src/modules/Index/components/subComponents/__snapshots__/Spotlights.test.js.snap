// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spotlights default values render correctly 1`] = `
<div>
  <div
    aria-label="UQ Spotlights carousel loading"
    data-testid="spotlights"
    role="region"
    style="height: 100%; position: relative; flex-grow: 1; border-radius: 4px; overflow: hidden;"
  >
    <svg
      aria-labelledby="personalisation-panel-or-hours-aria"
      height="100%"
      role="img"
      viewBox="0 0 1500 528"
      width="100%"
    >
      <title
        id="personalisation-panel-or-hours-aria"
      >
        Loading...
      </title>
      <rect
        clip-path="url(#personalisation-panel-or-hours-diff)"
        height="100%"
        role="presentation"
        style="fill: url(#personalisation-panel-or-hours-animated-diff);"
        width="100%"
        x="0"
        y="0"
      />
      <defs>
        <clippath
          id="personalisation-panel-or-hours-diff"
        >
          <rect
            height="721"
            rx="5"
            ry="5"
            width="1967"
            x="0"
            y="0"
          />
        </clippath>
        <lineargradient
          id="personalisation-panel-or-hours-animated-diff"
        >
          <stop
            offset="0%"
            stop-color="#f3f3f3"
            stop-opacity="1"
          >
            <animate
              attributeName="offset"
              dur="2s"
              keyTimes="0; 0.25; 1"
              repeatCount="indefinite"
              values="-2; -2; 1"
            />
          </stop>
          <stop
            offset="50%"
            stop-color="#e2e2e2"
            stop-opacity="1"
          >
            <animate
              attributeName="offset"
              dur="2s"
              keyTimes="0; 0.25; 1"
              repeatCount="indefinite"
              values="-1; -1; 2"
            />
          </stop>
          <stop
            offset="100%"
            stop-color="#f3f3f3"
            stop-opacity="1"
          >
            <animate
              attributeName="offset"
              dur="2s"
              keyTimes="0; 0.25; 1"
              repeatCount="indefinite"
              values="0; 0; 3"
            />
          </stop>
        </lineargradient>
      </defs>
    </svg>
  </div>
</div>
`;
